Wrangling and Assessing the Data

In [1]:
import pandas as pd
import datetime as dt
import numpy as np
import requests
import tweepy
import json
import re
from nltk import pos_tag
import time

Wrangling Data
1.Twitter Archive Data from csv file.
2.Image data from the given URL.
3.Tweets as JSON file from Twitter.

In [2]:
twitter_archive_df = pd.read_csv('twitter-archive-enhanced.csv')
twitter_archive_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [3]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

with open(url.split('/')[-1], mode='wb') as file:
    file.write(response.content)
    
images_df = pd.read_csv('image-predictions.tsv', sep='\t')
images_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [4]:
consumer_key = 'BICW51UxgQhzKTdPPfBsj2FCv'
consumer_secret = '6ISKYv7XK705wwp0eWtyYBB76JrZjhG9bxaeHXtvfOSBU6iwRa'
access_token = '1465192766-bv2SUlAELcvAz62qop3FjW3G04hY426GRLWMpnL'
access_token_secret = 'K7cW18Ci0wbxQjVvYkgjasoO7HUNx7cnFfxtjFXONPmM0'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)

# Construct the API instance
api = tweepy.API(auth, 
                 parser = tweepy.parsers.JSONParser(), # Parse the result to Json Object
                 wait_on_rate_limit = True, # Automatically wait for rate limits to replenish
                 wait_on_rate_limit_notify = True) # Print a notification when Tweepy is waiting for rate limits to replenish

In [5]:

# Liste where we will store the dictionaries of our result
'''df_list = []
# Liste frame where we will store the tweet_id of the errors
error_list = []

# Calculate the time of excution
start = time.time()

# Get the tweet object for all the teweets in archive dataframe 
for tweet_id in twitter_archive_df['tweet_id']:
    try:
        tweets = api.get_status(tweet_id, tweet_mode = 'extended')
        # Print one page to look at the structure of the returned file
        # and the names of attributes
        # print(json.dumps(page, indent = 4))
        #break
        
        favorites = tweets['favorite_count'] # How many favorites the tweet had
        retweets = tweets['retweet_count'] # Count of the retweet
        user_followers = tweets['user']['followers_count'] # How many followers the user had
        user_favourites = tweets['user']['favourites_count'] # How many favorites the user had
        date_time = tweets['created_at'] # The date and time of the creation
        
        df_list.append({'tweet_id': int(tweet_id),
                        'favorites': int(favorites),
                        'retweets': int(retweets),
                        'user_followers': int(user_followers),
                        'user_favourites': int(user_favourites),
                        'date_time': pd.to_datetime(date_time)})
    
    # Catch the exceptions of the TweepError
    except Exception as e:
        print(str(tweet_id)+ " _ " + str(e))
        error_list.append(tweet_id)

# Calculate the time of excution
end = time.time()
print(end - start)
'''

'df_list = []\n# Liste frame where we will store the tweet_id of the errors\nerror_list = []\n\n# Calculate the time of excution\nstart = time.time()\n\n# Get the tweet object for all the teweets in archive dataframe \nfor tweet_id in twitter_archive_df[\'tweet_id\']:\n    try:\n        tweets = api.get_status(tweet_id, tweet_mode = \'extended\')\n        # Print one page to look at the structure of the returned file\n        # and the names of attributes\n        # print(json.dumps(page, indent = 4))\n        #break\n        \n        favorites = tweets[\'favorite_count\'] # How many favorites the tweet had\n        retweets = tweets[\'retweet_count\'] # Count of the retweet\n        user_followers = tweets[\'user\'][\'followers_count\'] # How many followers the user had\n        user_favourites = tweets[\'user\'][\'favourites_count\'] # How many favorites the user had\n        date_time = tweets[\'created_at\'] # The date and time of the creation\n        \n        df_list.append({\'

In [6]:
#json_tweets = pd.DataFrame(df_list, columns = ['tweet_id', 'favorites', 'retweets','user_followers', 'user_favourites', 'date_time'])

# Save the dataFrame in file
#json_tweets.to_csv('tweet_json.txt', encoding = 'utf-8', index=False)

In [7]:
json_tweets_df = pd.read_csv('tweet_json.txt', encoding = 'utf-8')
json_tweets_df.head()

tweet_id  favorites  retweets  user_followers  user_favourites  \
0  892420643555336193      38331      8415         7336654           138622   
1  892177421306343426      32850      6203         7336653           138622   
2  891815181378084864      24743      4104         7336653           138622   
3  891689557279858688      41675      8542         7336653           138622   
4  891327558926688256      39847      9268         7336653           138622   

             date_time  
0  2017-08-01 16:23:56  
1  2017-08-01 00:17:27  
2  2017-07-31 00:18:03  
3  2017-07-30 15:58:51  
4  2017-07-29 16:00:24

Assesing all the dataframes:


In [8]:
twitter_archive_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [9]:
twitter_archive_df.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

Max value of denominator is 176 and numerator is 1776 which looks like unusual.

In [10]:
len(twitter_archive_df),len(images_df),len(json_tweets_df)

(2356, 2075, 2341)

In [11]:
twitter_archive_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [12]:
twitter_archive_df[twitter_archive_df['rating_denominator'] < 10].tweet_id.count()


3

In [13]:
twitter_archive_df[twitter_archive_df['rating_numerator'] > 50].tweet_id.count()

18

In [14]:
twitter_archive_df[twitter_archive_df['name'].apply(len) < 3].tweet_id.count()


79

In [15]:
twitter_archive_df['name'].value_counts()

None         745
a             55
Charlie       12
Cooper        11
Oliver        11
Lucy          11
Lola          10
Tucker        10
Penny         10
Bo             9
Winston        9
Sadie          8
the            8
Buddy          7
an             7
Bailey         7
Daisy          7
Toby           7
Jax            6
Scout          6
Koda           6
Dave           6
Stanley        6
Leo            6
Rusty          6
Oscar          6
Bella          6
Jack           6
Milo           6
George         5
            ... 
Cora           1
General        1
Lolo           1
Kona           1
Naphaniel      1
Grey           1
Sully          1
Flash          1
Rodman         1
Derby          1
Kayla          1
Remus          1
Ralf           1
Vince          1
Ginger         1
Julius         1
Opie           1
Grizzwald      1
Halo           1
Rumpole        1
Petrick        1
Miguel         1
Skittle        1
Harnold        1
Hector         1
River          1
Mason          1
Bode          

In [16]:
twitter_archive_df[twitter_archive_df['text'].isnull()]


Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [17]:
twitter_archive_df[twitter_archive_df['tweet_id'].isnull()]


Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

Number of retweets : 181

In [18]:
twitter_archive_df[twitter_archive_df['retweeted_status_id'].notnull()].tweet_id.count()



181

In [19]:
twitter_archive_df[twitter_archive_df.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [20]:
twitter_archive_df[twitter_archive_df.text.str.contains('&amp;')]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
262   842765311967449089                    NaN                  NaN   
273   840728873075638272                    NaN                  NaN   
320   834458053273591808                    NaN                  NaN   
461   817536400337801217                    NaN                  NaN   
485   814578408554463233                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
799   772826264096874500                    NaN                  NaN   
889   759793422261743616                    NaN                  NaN   
898   758854675097526272                    NaN                  NaN   
976   750026558547456000                    NaN                  NaN   
1104  735137028879360001                    NaN                  NaN   
1179  719367763014393856                    NaN                  NaN   
1199  716791146589110272                    NaN                  NaN   
1222  714258258790387713                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1366  702671118226825216                    NaN                  NaN   
1421  698195409219559425                    NaN                  NaN   
1465  694352839993344000                    NaN                  NaN   
1481  693280720173801472                    NaN                  NaN   
1508  691483041324204033                    NaN                  NaN   
1524  690597161306841088                    NaN                  NaN   
1538  689835978131935233                    NaN                  NaN   
1569  687807801670897665                    NaN                  NaN   
1593  686386521809772549                    NaN                  NaN   
1621  684926975086034944                    NaN                  NaN   
1646  683834909291606017                    NaN                  NaN   
1707  680801747103793152                    NaN                  NaN   
1763  678446151570427904                    NaN                  NaN   
1795  677314812125323265                    NaN                  NaN   
1812  676811746707918848                    NaN                  NaN   
1817  676603393314578432                    NaN                  NaN   
1842  675870721063669760           6.757073e+17         4.196984e+09   
1897  674737130913071104                    NaN                  NaN   
1899  674670581682434048                    NaN                  NaN   
1901  674646392044941312                    NaN                  NaN   
1913  674372068062928900                    NaN                  NaN   
1931  674036086168010753                    NaN                  NaN   
2031  671768281401958400                    NaN                  NaN   
2037  671561002136281088                    NaN                  NaN   
2064  671154572044468225                    NaN                  NaN   
2084  670807719151067136                    NaN                  NaN   
2096  670755717859713024                    NaN                  NaN   
2137  670046952931721218                    NaN                  NaN   
2177  669037058363662336                    NaN                  NaN   
2190  668960084974809088                    NaN                  NaN   
2196  668852170888998912                    NaN                  NaN   
2207  668627278264475648                    NaN                  NaN   
2210  668620235289837568                    NaN                  NaN   
2216  668537837512433665                    NaN                  NaN   
2232  668221241640230912                    NaN                  NaN   
2246  667878741721415682                    NaN                  NaN   
2268  667517642048163840                    NaN                  NaN   
2293  667152164079423490                    NaN                  NaN   
2306  666835007768551424                    NaN                  NaN   

                      timestamp  \
262   2017-03-17 15:

In [21]:
minimum_category_df =twitter_archive_df[(twitter_archive_df.doggo != 'None') | 
                                          (twitter_archive_df.floofer != 'None')|
                                          (twitter_archive_df.pupper != 'None') | 
                                          (twitter_archive_df.puppo != 'None')].tweet_id.count()
minimum_category_df

380

In [22]:
no_category_df =twitter_archive_df[(twitter_archive_df.doggo == 'None') & 
                                          (twitter_archive_df.floofer == 'None')&
                                          (twitter_archive_df.pupper == 'None') &
                                          (twitter_archive_df.puppo == 'None')].tweet_id.count()
no_category_df

1976

In [23]:
images_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [24]:
images_df.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [25]:
images_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [26]:
no_pred = images_df[(images_df.p1_dog != True)|
                               (images_df.p2_dog != True)|
                               (images_df.p3_dog != True)]
no_pred

tweet_id                                            jpg_url  \
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
33    666430724426358785    https://pbs.twimg.com/media/CT-jNYqW4AAPi2M.jpg   
34    666435652385423360    https://pbs.twimg.com/media/CT-nsTQWEAEkyDn.jpg   
43    666776908487630848    https://pbs.twimg.com/media/CUDeDoWUYAAD-EM.jpg   
45    666786068205871104    https://pbs.twimg.com/media/CUDmZIkWcAAIPPe.jpg   
48    666826780179869698    https://pbs.twimg.com/media/CUELa0NUkAAscGC.jpg   
49    666835007768551424    https://pbs.twimg.com/media/CUES51dXIAEahyG.jpg   
50    666837028449972224    https://pbs.twimg.com/media/CUEUva1WsAA2jPb.jpg   
51    666983947667116034    https://pbs.twimg.com/media/CUGaXDhW4AY9JUH.jpg   
52    666996132027977728    https://pbs.twimg.com/media/CUGlb6iUwAITEbW.jpg   
53    667012601033924608    https://pbs.twimg.com/media/CUG0bC0U8AAw2su.jpg   
54    667044094246576128    https://pbs.twimg.com/media/CUHREBXXAAE6A9b.jpg   
56    667065535570550784    https://pbs.twimg.com/media/CUHkkJpXIAA2w3n.jpg   
69    667188689915760640    https://pbs.twimg.com/media/CUJUk2iWUAAVtOv.jpg   
73    667369227918143488    https://pbs.twimg.com/media/CUL4xR9UkAEdlJ6.jpg   
77    667437278097252352    https://pbs.twimg.com/media/CUM2qWaWoAUZ06L.jpg   
78    667443425659232256    https://pbs.twimg.com/media/CUM8QZwW4AAVsBl.jpg   
82    667491009379606528    https://pbs.twimg.com/media/CUNniSlUYAEj1Jl.jpg   
83    667495797102141441    https://pbs.twimg.com/media/CUNr4-7UwAAg2lq.jpg   
87    667524857454854144    https://pbs.twimg.com/media/CUOGUfJW4AA_eni.jpg   
92    667546741521195010    https://pbs.twimg.com/media/CUOaOWXWcAA0_Jy.jpg   
93    667549055577362432    https://pbs.twimg.com/media/CUOcVCwWsAERUKY.jpg   
...                  ...                                                ...   
1972  869596645499047938    https://pbs.twimg.com/media/DBFtiYqWAAAsjj1.jpg   
1975  870063196459192321    https://pbs.twimg.com/media/DBMV3NnXUAAm0Pp.jpg   
1979  870804317367881728    https://pbs.twimg.com/media/DBW35ZsVoAEWZUU.jpg   
1980  871032628920680449    https://pbs.twimg.com/media/DBaHi3YXgAE6knM.jpg   
1981  871515927908634625    https://pbs.twimg.com/media/DBg_HT9WAAEeIMM.jpg   
1984  872122724285648897    https://pbs.twimg.com/media/DBpm-5UXcAUeCru.jpg   
1992  873697596434513921    https://pbs.twimg.com/media/DA7iHL5U0AA1OQo.jpg   
1998  875144289856114688  https://pbs.twimg.com/ext_tw_video_thumb/87514...   
2005  877556246731214848    https://pbs.twimg.com/media/DC20wEcW0AAf59m.jpg   
2012  879050749262655488    https://pbs.twimg.com/media/DDMD_phXoAQ1qf0.jpg   
2013  879376492567855104    https://pbs.twimg.com/media/DDQsQGFV0AAw6u9.jpg   
2015  879492040517615616    https://pbs.twimg.com/media/DDSVWMvXsAEgmMK.jpg   
2018  880221127280381952    https://pbs.twimg.com/media/DDcscbXU0AIfDzs.jpg   
2021  880935762899988482    https://pbs.twimg.com/media/DDm2Z5aXUAEDS2u.jpg   
2022  881268444196462592    https://pbs.twimg.com/media/DDrk-f9WAAI-WQv.jpg   
2023  881536004380872706  https://pbs.twimg.com/ext_tw_video_thumb/88153...   
2025  881906580714921986    https://pbs.twimg.com/media/DD0pWm9XcAAeSBL.jpg   
2026  882045870035918850    https://pbs.twimg.com/media/DD2oCl2WAAEI_4a.jpg   
2030  882992080364220416    https://pbs.twimg.com/media/DEEEnIqXYA

In [27]:
json_tweets_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2341 entries, 0 to 2340
Data columns (total 6 columns):
tweet_id           2341 non-null int64
favorites          2341 non-null int64
retweets           2341 non-null int64
user_followers     2341 non-null int64
user_favourites    2341 non-null int64
date_time          2341 non-null object
dtypes: int64(5), object(1)
memory usage: 109.8+ KB


In [28]:
json_tweets_df.describe()

tweet_id      favorites      retweets  user_followers  \
count  2.341000e+03    2341.000000   2341.000000    2.341000e+03   
mean   7.422089e+17    8022.158906   2964.145237    7.336718e+06   
std    6.833609e+16   12383.027590   4975.364522    7.006652e+01   
min    6.660209e+17       0.000000      0.000000    7.336652e+06   
25%    6.783411e+17    1389.000000    595.000000    7.336657e+06   
50%    7.186133e+17    3499.000000   1384.000000    7.336715e+06   
75%    7.986979e+17    9847.000000   3456.000000    7.336768e+06   
max    8.924206e+17  165065.000000  84440.000000    7.337448e+06   

       user_favourites  
count      2341.000000  
mean     138622.855617  
std           0.388509  
min      138620.000000  
25%      138623.000000  
50%      138623.000000  
75%      138623.000000  
max      138624.000000

In [29]:
json_tweets_df.head()

tweet_id  favorites  retweets  user_followers  user_favourites  \
0  892420643555336193      38331      8415         7336654           138622   
1  892177421306343426      32850      6203         7336653           138622   
2  891815181378084864      24743      4104         7336653           138622   
3  891689557279858688      41675      8542         7336653           138622   
4  891327558926688256      39847      9268         7336653           138622   

             date_time  
0  2017-08-01 16:23:56  
1  2017-08-01 00:17:27  
2  2017-07-31 00:18:03  
3  2017-07-30 15:58:51  
4  2017-07-29 16:00:24

In [30]:
json_tweets_df[json_tweets_df.date_time > '2017-06-29 16:00:00']


tweet_id  favorites  retweets  user_followers  user_favourites  \
0   892420643555336193      38331      8415         7336654           138622   
1   892177421306343426      32850      6203         7336653           138622   
2   891815181378084864      24743      4104         7336653           138622   
3   891689557279858688      41675      8542         7336653           138622   
4   891327558926688256      39847      9268         7336653           138622   
5   891087950875897856      19999      3077         7336653           138622   
6   890971913173991426      11706      2043         7336653           138622   
7   890729181411237888      64707     18656         7336653           138622   
8   890609185150312448      27505      4218         7336653           138622   
9   890240255349198849      31553      7304         7336653           138622   
10  890006608113172480      30314      7246         7336653           138622   
11  889880896479866881      27470      4921         7336653           138622   
12  889665388333682689      47580      9945         7336653           138622   
13  889638837579907072      26861      4492         7336653           138622   
14  889531135344209921      14921      2213         7336653           138622   
15  889278841981685760      25013      5357         7336653           138622   
16  888917238123831296      28768      4450         7336653           138622   
17  888804989199671297      25296      4274         7336653           138622   
18  888554962724278272      19654      3534         7336653           138622   
19  888078434458587136      21515      3456         7336653           138622   
20  887705289381826560      29822      5327         7336653           138622   
21  887517139158093824      45735     11532         7336653           138622   
22  887473957103951883      68303     17991         7336653           138622   
23  887343217045368832      33296     10290         7336653           138622   
24  887101392804085760      30216      5896         7336653           138622   
25  886983233522544640      34759      7691         7336653           138622   
26  886736880519319552      11916      3251         7336653           138622   
27  886680336477933568      22175      4414         7336653           138622   
28  886366144734445568      20956      3166         7336653           138622   
29  886267009285017600        116         4         7336653           138622   
30  886258384151887873      27672      6209         7336653           138622   
31  886054160059072513          0       104         7336653           138622   
32  885984800019947520      32316      6736         7337161           138623   
33  885528943205470208      35589      6362         7336653           138622   
34  885518971528720385      20201      3712         7336653           138622   
35  885311592912609280          0     18394         7336653           138622   
36  885167619883638784      21676      4336         7336653           138622   
37  884925521741709313      76616     18239         7336653           138622   
38  884876753390489601      27607      5728         7336653           138622   
39  884562892145688576      24041      4817         7336653           138622   
40  884441805382717440      26690      5592         7336653           138622   
41  884247878851493888      72232     20251         7336653           138622   
42  884162670584377345      20137      2953         7336653           138622   
43  883838122936631299      21658      3410         7336653           138622   
44  883482846933004288      45437      9858         7336653           138622   
45  883360690899218434      22327      3641         7336653           138622   
46  883117836046086144      36784      6609         7336653           138622   
47  882992080364220416      23696      3878         7336653           138622   
48  882762694511734784      28001      4869         7336653           138622   
49  8826272703

In [31]:
json_tweets_df[json_tweets_df.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, favorites, retweets, user_followers, user_favourites, date_time]
Index: []

In [32]:
all_columns = pd.Series(list(twitter_archive_df) + list(json_tweets_df) + list(images_df))
all_columns[all_columns.duplicated()]

17    tweet_id
23    tweet_id
dtype: object

Quality Issues:

(accuracy, validity, consistency, completeness)

1. Need to remove rows that have been retweeted, they are not original tweets-column name:retweets-181 records.
2. Retweets associated columns not needed (only need original tweets with images). Those irrelevant columns are 'in_reply_to_user_id', 'in_reply_to_status_id', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'.
3. Timestamp is string instead of timestamp.
4. Several ratings are simply done wrong i.e The numerator and denominator columns have too large/too small values(We know the numerator could be large but too big denominator is not acceptable)
5. Columns doggo, floofer, pupper, and puppo have value None. This should be nulls.
6. The name column has a lot of non-name values. The most popular name is 'a' which is not a name at all.
7. tweet_id should be string.
8. Column name for each type of dog. Should be 1 column and 4 categorical choices.
9. The timestamp type is an object, not a timestamp.



image_predictions¶

1. There were 2075 rows in the images dataframe compared to 2356 rows in the archive dataframe. This due to non-pictures and retweets included.
2. Column names are not informative.
3. tweet_id should be string.
4. Capitalize breed names.
5. Change breed to categorical.

twitter_counts
1. tweet_id is numeric. Should be string.












Tidiness Issues:

1. Dog stage(doggo, floofer, pupper, puppo) in different columns.
2. Three separate tables(dataframes)-We only need one table that is a subset of these 3 tables.
3. retweets and favorites in two tables.
4. There are 2075 rows in prediction_images and 2356 rows in twitter_archive.
5. Drop columns that are not needed & rearrange column order for an easier read





***Cleaning Data***
<br>
Making copies of all three dataframes

In [33]:
twitter_archive_clean_df =twitter_archive_df.copy()
images_clean_df = images_df.copy()
json_tweets_clean_df = json_tweets_df.copy()

Merging the data in one dataframe

In [34]:
master_df = pd.merge(twitter_archive_clean_df, json_tweets_clean_df,
                        how = 'left', on = 'tweet_id')
master_df = pd.merge(master_df, images_clean_df,
                        how = 'left', on = 'tweet_id')
master_df.to_csv('master_df.csv',encoding='utf-8')

In [35]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 33 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
favorites                     23

In [37]:
#drop duplicates
master_df.drop_duplicates(inplace=True)

In [38]:
# Delete tweets with no pictures
master_df = master_df.dropna(subset = ['jpg_url'])




In [39]:
master_df = master_df[pd.isnull(master_df.retweeted_status_id)]


In [40]:
master_df.shape

(1994, 33)

In [41]:
del_columns = ['retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp','date_time']
master_df.drop(del_columns,axis=1,inplace=True)

In [42]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 2355
Data columns (total 29 columns):
tweet_id                 1994 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                1994 non-null object
source                   1994 non-null object
text                     1994 non-null object
expanded_urls            1994 non-null object
rating_numerator         1994 non-null int64
rating_denominator       1994 non-null int64
name                     1994 non-null object
doggo                    1994 non-null object
floofer                  1994 non-null object
pupper                   1994 non-null object
puppo                    1994 non-null object
favorites                1993 non-null float64
retweets                 1993 non-null float64
user_followers           1993 non-null float64
user_favourites          1993 non-null float64
jpg_url                  1994 non-null object
img_num             

In [43]:
columns_to_melt = ['doggo', 'floofer', 'pupper', 'puppo']
columns_to_stay = [x for x in master_df.columns.tolist() if x not in columns_to_melt]
master_df = pd.melt(master_df, id_vars = columns_to_stay, value_vars = columns_to_melt, 
                         var_name = 'stages', value_name = 'dog_stage')
master_df = master_df.drop('stages', axis=1)



In [44]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7976 entries, 0 to 7975
Data columns (total 26 columns):
tweet_id                 7976 non-null int64
in_reply_to_status_id    92 non-null float64
in_reply_to_user_id      92 non-null float64
timestamp                7976 non-null object
source                   7976 non-null object
text                     7976 non-null object
expanded_urls            7976 non-null object
rating_numerator         7976 non-null int64
rating_denominator       7976 non-null int64
name                     7976 non-null object
favorites                7972 non-null float64
retweets                 7972 non-null float64
user_followers           7972 non-null float64
user_favourites          7972 non-null float64
jpg_url                  7976 non-null object
img_num                  7976 non-null float64
p1                       7976 non-null object
p1_conf                  7976 non-null float64
p1_dog                   7976 non-null object
p2                

In [45]:
print(master_df.dog_stage.value_counts())

None       7659
pupper      212
doggo        74
puppo        23
floofer       8
Name: dog_stage, dtype: int64


In [46]:
master_df[master_df.tweet_id.duplicated()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1994  892420643555336193                    NaN                  NaN   
1995  892177421306343426                    NaN                  NaN   
1996  891815181378084864                    NaN                  NaN   
1997  891689557279858688                    NaN                  NaN   
1998  891327558926688256                    NaN                  NaN   
1999  891087950875897856                    NaN                  NaN   
2000  890971913173991426                    NaN                  NaN   
2001  890729181411237888                    NaN                  NaN   
2002  890609185150312448                    NaN                  NaN   
2003  890240255349198849                    NaN                  NaN   
2004  890006608113172480                    NaN                  NaN   
2005  889880896479866881                    NaN                  NaN   
2006  889665388333682689                    NaN                  NaN   
2007  889638837579907072                    NaN                  NaN   
2008  889531135344209921                    NaN                  NaN   
2009  889278841981685760                    NaN                  NaN   
2010  888917238123831296                    NaN                  NaN   
2011  888804989199671297                    NaN                  NaN   
2012  888554962724278272                    NaN                  NaN   
2013  888078434458587136                    NaN                  NaN   
2014  887705289381826560                    NaN                  NaN   
2015  887517139158093824                    NaN                  NaN   
2016  887473957103951883                    NaN                  NaN   
2017  887343217045368832                    NaN                  NaN   
2018  887101392804085760                    NaN                  NaN   
2019  886983233522544640                    NaN                  NaN   
2020  886736880519319552                    NaN                  NaN   
2021  886680336477933568                    NaN                  NaN   
2022  886366144734445568                    NaN                  NaN   
2023  886258384151887873                    NaN                  NaN   
...                  ...                    ...                  ...   
7946  666411507551481857                    NaN                  NaN   
7947  666407126856765440                    NaN                  NaN   
7948  666396247373291520                    NaN                  NaN   
7949  666373753744588802                    NaN                  NaN   
7950  666362758909284353                    NaN                  NaN   
7951  666353288456101888                    NaN                  NaN   
7952  666345417576210432                    NaN                  NaN   
7953  666337882303524864                    NaN                  NaN   
7954  666293911632134144                    NaN                  NaN   
7955  666287406224695296                    NaN                  NaN   
7956  666273097616637952                    NaN                  NaN   
7957  666268910803644416                    NaN                  NaN   
7958  666104133288665088                    NaN                  NaN   
7959  666102155909144576                    NaN                  NaN   
7960  666099513787052032                    NaN                  NaN   
7961  666094000022159362                    NaN                  NaN   
7962  666082916733198337                    NaN                  NaN   
7963  666073100786774016                    NaN                  NaN   
7964  666071193221509120                    NaN                  NaN   
7965  666063827256086533                    NaN                  NaN   
7966  666058600524156928                    NaN                  NaN   
7967  666057090499244032                    NaN                  NaN   
7968  666055525042405380                    NaN                  NaN   
7969  666051853826850816                    NaN         

In [47]:
master_df.drop_duplicates(inplace=True)

In [64]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2311 entries, 0 to 6795
Data columns (total 15 columns):
tweet_id                2311 non-null object
timestamp               2311 non-null datetime64[ns]
source                  2311 non-null object
text                    2311 non-null object
expanded_urls           2311 non-null object
rating_numerator        2311 non-null float64
rating_denominator      2311 non-null float64
name                    2311 non-null object
favorites               2310 non-null float64
retweets                2310 non-null float64
user_followers          2310 non-null float64
jpg_url                 2311 non-null object
dog_stage               2311 non-null object
prediction_algorithm    2311 non-null object
confidence_level        2311 non-null float64
dtypes: datetime64[ns](1), float64(6), object(8)
memory usage: 288.9+ KB


In [65]:
master_df.tweet_id = master_df.tweet_id.astype(str)
master_df['rating_numerator'] = master_df['rating_numerator'].astype(float)
master_df['rating_denominator'] = master_df['rating_denominator'].astype(float)
master_df.timestamp = pd.to_datetime(master_df.timestamp)




Get rid of image prediction columns¶


In [49]:
# We will store the fisrt true algorithm with it's level of confidence
prediction_algorithm = []
confidence_level = []

# Get_prediction_confidence function:
# search the first true algorithm and append it to a list with it's level of confidence
# if flase prediction_algorthm will have a value of NaN
def get_prediction_confidence(dataframe):
    if dataframe['p1_dog'] == True:
        prediction_algorithm.append(dataframe['p1'])
        confidence_level.append(dataframe['p1_conf'])
    elif dataframe['p2_dog'] == True:
        prediction_algorithm.append(dataframe['p2'])
        confidence_level.append(dataframe['p2_conf'])
    elif dataframe['p3_dog'] == True:
        prediction_algorithm.append(dataframe['p3'])
        confidence_level.append(dataframe['p3_conf'])
    else:
        prediction_algorithm.append('NaN')
        confidence_level.append(0)

master_df.apply(get_prediction_confidence, axis=1)
master_df['prediction_algorithm'] = prediction_algorithm
master_df['confidence_level'] = confidence_level

# Test
list(master_df)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'favorites',
 'retweets',
 'user_followers',
 'user_favourites',
 'jpg_url',
 'img_num',
 'p1',
 'p1_conf',
 'p1_dog',
 'p2',
 'p2_conf',
 'p2_dog',
 'p3',
 'p3_conf',
 'p3_dog',
 'dog_stage',
 'prediction_algorithm',
 'confidence_level']

In [50]:
master_df = master_df.drop(['img_num', 'p1', 'p1_conf', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog'], 1)

# Test
list(master_df)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'favorites',
 'retweets',
 'user_followers',
 'user_favourites',
 'jpg_url',
 'dog_stage',
 'prediction_algorithm',
 'confidence_level']

In [142]:
master_df.drop_duplicates(inplace=True)

In [53]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2311 entries, 0 to 6795
Data columns (total 18 columns):
tweet_id                 2311 non-null object
in_reply_to_status_id    27 non-null float64
in_reply_to_user_id      27 non-null float64
timestamp                2311 non-null datetime64[ns]
source                   2311 non-null object
text                     2311 non-null object
expanded_urls            2311 non-null object
rating_numerator         2311 non-null float64
rating_denominator       2311 non-null float64
name                     2311 non-null object
favorites                2310 non-null float64
retweets                 2310 non-null float64
user_followers           2310 non-null float64
user_favourites          2310 non-null float64
jpg_url                  2311 non-null object
dog_stage                2311 non-null object
prediction_algorithm     2311 non-null object
confidence_level         2311 non-null float64
dtypes: datetime64[ns](1), float64(9), object(8)
mem

In [57]:
# let's concentrate on low values.. let's dig more
master_df['in_reply_to_user_id'].value_counts()
#since all the id's are same,hence we can delete this column as well

4.196984e+09    27
Name: in_reply_to_user_id, dtype: int64

In [58]:
master_df['source'].value_counts()


<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2269
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       29
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      13
Name: source, dtype: int64

In [59]:
master_df.source = master_df.source.str.replace('<','>').str.split('>').str[2]



In [60]:
master_df['source'].value_counts()



Twitter for iPhone    2269
Twitter Web Client      29
TweetDeck               13
Name: source, dtype: int64

In [61]:
master_df['user_favourites'].value_counts()

138623.0    2009
138622.0     284
138624.0       9
138620.0       8
Name: user_favourites, dtype: int64

In [62]:
master_df = master_df.drop(['in_reply_to_status_id', 'in_reply_to_user_id', 'user_favourites'], axis=1)

In [66]:
male = ['He', 'he', 'him', 'his', "he's", 'himself']
female = ['She', 'she', 'her', 'hers', 'herself', "she's"]

dog_gender = []

for text in master_df['text']:
    # Male
    if any(map(lambda v:v in male, text.split())):
        dog_gender.append('male')
    # Female
    elif any(map(lambda v:v in female, text.split())):
        dog_gender.append('female')
    # If group or not specified
    else:
        dog_gender.append('NaN')

# Test
len(dog_gender)

# Save the result in a new column 'dog_name'
master_df['dog_gender'] = dog_gender

In [67]:
master_df.dog_gender.value_counts()

NaN       1325
male       722
female     264
Name: dog_gender, dtype: int64

In [68]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2311 entries, 0 to 6795
Data columns (total 16 columns):
tweet_id                2311 non-null object
timestamp               2311 non-null datetime64[ns]
source                  2311 non-null object
text                    2311 non-null object
expanded_urls           2311 non-null object
rating_numerator        2311 non-null float64
rating_denominator      2311 non-null float64
name                    2311 non-null object
favorites               2310 non-null float64
retweets                2310 non-null float64
user_followers          2310 non-null float64
jpg_url                 2311 non-null object
dog_stage               2311 non-null object
prediction_algorithm    2311 non-null object
confidence_level        2311 non-null float64
dog_gender              2311 non-null object
dtypes: datetime64[ns](1), float64(6), object(9)
memory usage: 306.9+ KB


In [76]:
master_df['dog_gender'] = master_df['dog_gender'].astype('category')
master_df['source'] = master_df['source'].astype('category')
master_df['dog_stage'] = master_df['dog_stage'].astype('category')

In [77]:
master_df = master_df.rename(columns = {'timestamp': 'tweet_date', 'source': 'tweet_source', 'text': 'tweet_text', 
                                        'expanded_urls': 'tweet_url', 'jpg_url': 'tweet_picture_predicted', 
                                        'favorites': 'tweet_favorites', 'retweets': 'tweet_retweets',
                                        'prediction_algorithm' : 'dog_breed'})

In [155]:
master_df.name.replace('None', np.nan, inplace = True)


In [158]:
master_df.name.value_counts()


a           59
Cooper      13
Oliver      12
Charlie     12
Lucy        11
Winston      9
Penny        9
Tucker       9
the          9
Bo           8
Sadie        8
Lola         7
Toby         7
Daisy        7
Jax          7
Milo         7
Bella        7
one          6
Maggie       6
an           6
Stanley      6
just         6
Gus          6
Scout        6
Koda         6
Cassie       5
Dave         5
Buddy        5
Scooter      5
Rusty        5
            ..
Hercules     1
Binky        1
Berkeley     1
Crawford     1
Oreo         1
Sky          1
Ralpher      1
Spencer      1
Nollie       1
Terrance     1
Margo        1
Gustav       1
Glacier      1
Eugene       1
Ben          1
Suki         1
Spanky       1
Rupert       1
Dug          1
Dante        1
Tess         1
JD           1
Derby        1
Nico         1
Brian        1
Sonny        1
Travis       1
Mack         1
Meyer        1
Dex          1
Name: name, Length: 935, dtype: int64